MainWindow.cpp

```diff
@@ -73,15 +77,44 @@ void diagnosticHandler(QtMsgType type, const QMessageLogContext &context, const
 
   std::cerr << msgFormatted.toStdString() << std::endl;
 
-  if (diagnosticTextEdit) {
+  if (diagnosticTextEdit != nullptr) {
     diagnosticTextEdit->append(msgFormatted);
   } else {
     // this should never happen!
     std::cerr << "Critical: Diagnostics text control does not exist!" << std::endl;
   }
 }
 
-MainWindow::MainWindow(QWidget *parent) : QMainWindow(parent), _ui(new Ui::MainWindow) {
+QFileInfo MainWindow::getEnigmaRoot() {
+  QFileInfo EnigmaRoot;
+  foreach (auto path, EnigmaSearchPaths) {
+    const QDir dir(path);
+    QDir::Filters filters = QDir::Filter::AllEntries;
+    auto entryList = dir.entryInfoList(QStringList({"ENIGMAsystem"}), filters, QDir::SortFlag::NoSort);
+    if (!entryList.empty()) {
+      EnigmaRoot = entryList.first();
+        break;
+    }
+  }
+  
+  return EnigmaRoot;
+}
+
+MainWindow::MainWindow(QWidget *parent) : QMainWindow(parent), _ui(new Ui::MainWindow), _event_data(nullptr), egm(nullptr) {
+
+  if (!EnigmaRoot.filePath().isEmpty()) {
+    _event_data = std::make_unique<EventData>(ParseEventFile((EnigmaRoot.absolutePath() + "/events.ey").toStdString()));
+  } else {
+    qDebug() << "Error: Failed to locate ENIGMA sources. Loading internal events.ey.\n" << "Search Paths:\n" << MainWindow::EnigmaSearchPaths;
+    QFile internal_events(":/events.ey");
+    internal_events.open(QIODevice::ReadOnly | QFile::Text);
+    std::stringstream ss;
+    ss << internal_events.readAll().toStdString();
+    _event_data = std::make_unique<EventData>(ParseEventFile(ss));
+  }
+  
+  egm = egm::EGM(_event_data.get());

```

Comment:
A lot of this should probably go into its own project handling class, like Josh had me do for recent files. Jamming everything into the main window class is not a good, we want to design this with loose coupling of components, something that signals and slots help accomplish.

Reply:
Sounds like a task for another day.

Reply:
This is new code, all _new_ code should be written correctly.

Reply:
This is just the first draft to make it work/build against enigma. As long as you hold up this pr over nonsense design choices the CI remains broken. This pr was meant to be merged the same time as the enigma side one. You can refactor it later but it needs to go in now.